### PR TITLE
AMD-related Code Changes

### PR DESCRIPTION
**Pull request by fthaler**
- Optimized boundary conditions kernel for GPUs, gives slightly better performance on NVIDIA GPUs and much better on AMD.
- Workaround for HIP-Clang compiler bug in GPU layout transformations: https://github.com/ROCm-Developer-Tools/HIP/issues/1283

### CODE REVIEW

**Comment by fthaler** [comment]
launch strgrid

**Comment by fthaler** [comment]
launch icgrid

**Comment by fthaler** [comment]
launch perftest

**Comment by fthaler** [comment]
launch icgrid

**Comment by fthaler** [comment]
launch strgrid

**Comment by fthaler** [comment]
launch perftest

File: include/gridtools/boundary_conditions/apply_gpu.hpp [hunk 1]
```diff
@@ -96,10 +96,8 @@ namespace gridtools {
                 array<uint_t, 3> m_start;
                 array<uint_t, 3> m_perm = {{0, 1, 2}};
 
-                GT_FUNCTION
```
Comment (havogt) [inline include/gridtools/boundary_conditions/apply_gpu.hpp]: Yeah! :1st_place_medal:

File: include/gridtools/boundary_conditions/apply_gpu.hpp [hunk 2]
```diff
@@ -185,96 +181,78 @@ namespace gridtools {
...
-                                configuration[1] = max(configuration[1], sizes[i][j][k].median());
-                                configuration[2] = max(configuration[2], sizes[i][j][k].min());
+                                b.x = std::max(b.x, (decltype(b.x))sizes[i][j][k].max());
+                                b.y = std::max(b.y, (decltype(b.y))sizes[i][j][k].median());
+                                b.z = std::max(b.z, (decltype(b.z))sizes[i][j][k].min());
                             }
                         }
                     }
                 }
+                assert((b.x > 0 || b.y > 0 || b.z > 0) && "all boundary extents are empty");
```
Comment (havogt) [inline include/gridtools/boundary_conditions/apply_gpu.hpp]: I don't fully understand what's going on here. Just wondering if in a generic context a boundary condition might be called with 0 extents and we want to silently do nothing?

File: include/gridtools/boundary_conditions/apply_gpu.hpp [hunk 3]
```diff
@@ -185,96 +181,78 @@ namespace gridtools {
...
-        GT_RUN_BC_ON(minus_, plus_, plus_);
-
-        GT_RUN_BC_ON(zero_, minus_, minus_);
-        GT_RUN_BC_ON(zero_, minus_, zero_);
-        GT_RUN_BC_ON(zero_, minus_, plus_);
-
-        GT_RUN_BC_ON(zero_, zero_, minus_);
-
-        GT_RUN_BC_ON(zero_, zero_, plus_);
+            dim3 blocks() const {
```
Comment (havogt) [inline include/gridtools/boundary_conditions/apply_gpu.hpp]: maybe rename to `required_blocks` or `compute_required_blocks`

File: include/gridtools/boundary_conditions/apply_gpu.hpp [hunk 4]
```diff
@@ -185,96 +181,78 @@ namespace gridtools {
...
+                dim3 b = block_size();
```
Comment (havogt) [inline include/gridtools/boundary_conditions/apply_gpu.hpp]: Is `block_size` the right name?

File: include/gridtools/interface/layout_transformation/layout_transformation_impl_cuda.hpp [hunk 5]
```diff
@@ -32,13 +35,13 @@ namespace gridtools {
             gridtools::array<gridtools::uint_t, GT_TRANSFORM_MAX_DIM> src_strides,
             gridtools::array<size_t, GT_TRANSFORM_MAX_DIM - 3> outer_dims) {
 
-            int i = blockIdx.x * blockDim.x + threadIdx.x;
+            int i = blockIdx.x * block_size_1d_t::value + threadIdx.x;
```
Comment (havogt) [inline include/gridtools/interface/layout_transformation/layout_transformation_impl_cuda.hpp]: comment why this workaround is needed

File: include/gridtools/boundary_conditions/apply_gpu.hpp (hunk 2 above)
Comment (fthaler) [reply include/gridtools/boundary_conditions/apply_gpu.hpp]: I have taken the assertion from @mbianco’s implementation (see line 332 of the file), no idea if this assertion can be ever triggered by a user!

Comment (havogt) [reply include/gridtools/boundary_conditions/apply_gpu.hpp]: Ok, I thought it is new.

File: include/gridtools/boundary_conditions/apply_gpu.hpp (hunk 4 above)
Comment (fthaler) [reply include/gridtools/boundary_conditions/apply_gpu.hpp]: Renamed, hope it’s more clear now.

File: include/gridtools/interface/layout_transformation/layout_transformation_impl_cuda.hpp (hunk 5 above)
Comment (fthaler) [reply include/gridtools/interface/layout_transformation/layout_transformation_impl_cuda.hpp]: Done.

File: include/gridtools/boundary_conditions/apply_gpu.hpp [hunk 6]
```diff
@@ -22,58 +27,53 @@ namespace gridtools {
...
+            GT_FUNCTION static constexpr uint_t direction_index(direction<I, J, K>) {
+                constexpr int_t stride_i = 9;
+                constexpr int_t stride_j = 3;
+                constexpr int_t stride_k = 1;
+                return (I + 1) * stride_i + (J + 1) * stride_j + (K + 1) * stride_k;
             }
+
+            GT_STATIC_ASSERT(sizeof(uint_t) >= 4, GT_INTERNAL_ERROR);
+
+            uint_t m_values;
```
[the remark this answers is not in the document]
Comment (fthaler) [reply include/gridtools/boundary_conditions/apply_gpu.hpp]: Done.

File: include/gridtools/boundary_conditions/apply_gpu.hpp (hunk 3 above)
Comment (fthaler) [reply include/gridtools/boundary_conditions/apply_gpu.hpp]: Renamed blocks() and threads_per_block(), hope it’s more clear now.

**Comment by fthaler** [comment]
launch strgrid

**Comment by fthaler** [comment]
launch perftest

**Pull request merged**